File: packages/snow_draw_arrow_core/lib/src/arrow_types.dart
import 'dart:math' as math;

typedef Point = List<double>;
typedef Bounds = List<double>;

typedef BindMode = String;
const String bindModeInside = 'inside';
const String bindModeOrbit = 'orbit';
const String bindModeSkip = 'skip';

typedef ArrowEndpointEdge = String;
typedef ArrowEndpointBindingField = String;
typedef ArrowEndpointSelector = String;

const String arrowEndpointStart = 'start';
const String arrowEndpointEnd = 'end';

String normalizeArrowEndpointEdge(ArrowEndpointSelector edge) =>
    edge == 'start' || edge == 'startBinding' ? 'start' : 'end';

const Map<String, int> bindableRoundness = {
  'LEGACY': 1,
  'PROPORTIONAL': 2,
  'ADAPTIVE': 3,
};

typedef BindableRoundnessType = Object;

class BindableRoundness {
  const BindableRoundness({required this.type, this.value});

  final BindableRoundnessType type;
  final double? value;
}

typedef Arrowhead = String;
typedef ArrowStrokeStyle = String;
typedef ArrowheadDashMode = String;
typedef ArrowheadFillMode = String;
typedef CanonicalBindableShape = String;
typedef BindableShape = String;

String canonicalizeBindableShape(BindableShape shape) {
  switch (shape) {
    case 'rect':
      return 'rectangle';
    case 'circle':
      return 'ellipse';
    case 'rhombus':
      return 'diamond';
    default:
      return shape;
  }
}

class CurvePathOp {
  const CurvePathOp({required this.op, required this.data});

  final String op;
  final List<double> data;
}

typedef ArrowheadPoints = List<double>;

class FixedPointBinding {
  const FixedPointBinding({
    required this.elementId,
    required this.fixedPoint,
    required this.mode,
  });

  final String elementId;
  final Point fixedPoint;
  final BindMode mode;

  FixedPointBinding copyWith({
    String? elementId,
    Point? fixedPoint,
    BindMode? mode,
  }) => FixedPointBinding(
    elementId: elementId ?? this.elementId,
    fixedPoint: fixedPoint ?? this.fixedPoint,
    mode: mode ?? this.mode,
  );
}

class FixedSegment {
  const FixedSegment({
    required this.start,
    required this.end,
    required this.index,
  });

  final Point start;
  final Point end;
  final int index;

  FixedSegment copyWith({Point? start, Point? end, int? index}) => FixedSegment(
    start: start ?? this.start,
    end: end ?? this.end,
    index: index ?? this.index,
  );
}

class BindableState {
  const BindableState({
    required this.id,
    required this.shape,
    required this.x,
    required this.y,
    required this.width,
    required this.height,
    required this.angle,
    required this.strokeWidth,
    this.roundness,
    this.zIndex,
    this.backgroundOpaque,
    this.bindingEnabled,
    this.interiorHitEnabled,
    this.visibilityBounds,
  });

  final String id;
  final BindableShape shape;
  final double x;
  final double y;
  final double width;
  final double height;
  final double angle;
  final double strokeWidth;
  final BindableRoundness? roundness;
  final double? zIndex;
  final bool? backgroundOpaque;
  final bool? bindingEnabled;
  final bool? interiorHitEnabled;
  final Bounds? visibilityBounds;

  BindableState copyWith({
    String? id,
    BindableShape? shape,
    double? x,
    double? y,
    double? width,
    double? height,
    double? angle,
    double? strokeWidth,
    BindableRoundness? roundness,
    double? zIndex,
    bool? backgroundOpaque,
    bool? bindingEnabled,
    bool? interiorHitEnabled,
    Bounds? visibilityBounds,
    bool setVisibilityBounds = false,
  }) => BindableState(
    id: id ?? this.id,
    shape: shape ?? this.shape,
    x: x ?? this.x,
    y: y ?? this.y,
    width: width ?? this.width,
    height: height ?? this.height,
    angle: angle ?? this.angle,
    strokeWidth: strokeWidth ?? this.strokeWidth,
    roundness: roundness ?? this.roundness,
    zIndex: zIndex ?? this.zIndex,
    backgroundOpaque: backgroundOpaque ?? this.backgroundOpaque,
    bindingEnabled: bindingEnabled ?? this.bindingEnabled,
    interiorHitEnabled: interiorHitEnabled ?? this.interiorHitEnabled,
    visibilityBounds: setVisibilityBounds
        ? visibilityBounds
        : this.visibilityBounds,
  );
}

class ArrowState {
  const ArrowState({
    required this.id,
    required this.x,
    required this.y,
    required this.width,
    required this.height,
    required this.points,
    required this.startBinding,
    required this.endBinding,
    required this.startArrowhead,
    required this.endArrowhead,
    required this.elbowed,
    required this.fixedSegments,
    required this.startIsSpecial,
    required this.endIsSpecial,
  });

  final String id;
  final double x;
  final double y;
  final double width;
  final double height;
  final List<Point> points;
  final FixedPointBinding? startBinding;
  final FixedPointBinding? endBinding;
  final Arrowhead? startArrowhead;
  final Arrowhead? endArrowhead;
  final bool elbowed;
  final List<FixedSegment>? fixedSegments;
  final bool? startIsSpecial;
  final bool? endIsSpecial;

  ArrowState copyWith({
    String? id,
    double? x,
    double? y,
    double? width,
    double? height,
    List<Point>? points,
    FixedPointBinding? startBinding,
    bool setStartBinding = false,
    FixedPointBinding? endBinding,
    bool setEndBinding = false,
    Arrowhead? startArrowhead,
    bool setStartArrowhead = false,
    Arrowhead? endArrowhead,
    bool setEndArrowhead = false,
    bool? elbowed,
    List<FixedSegment>? fixedSegments,
    bool setFixedSegments = false,
    bool? startIsSpecial,
    bool setStartIsSpecial = false,
    bool? endIsSpecial,
    bool setEndIsSpecial = false,
  }) => ArrowState(
    id: id ?? this.id,
    x: x ?? this.x,
    y: y ?? this.y,
    width: width ?? this.width,
    height: height ?? this.height,
    points: points ?? this.points,
    startBinding: setStartBinding ? startBinding : this.startBinding,
    endBinding: setEndBinding ? endBinding : this.endBinding,
    startArrowhead: setStartArrowhead ? startArrowhead : this.startArrowhead,
    endArrowhead: setEndArrowhead ? endArrowhead : this.endArrowhead,
    elbowed: elbowed ?? this.elbowed,
    fixedSegments: setFixedSegments ? fixedSegments : this.fixedSegments,
    startIsSpecial: setStartIsSpecial ? startIsSpecial : this.startIsSpecial,
    endIsSpecial: setEndIsSpecial ? endIsSpecial : this.endIsSpecial,
  );
}

class EngineContext {
  const EngineContext({
    required this.zoom,
    required this.isBindingEnabled,
    required this.bindMode,
    required this.maxCoordinate,
  });

  final double zoom;
  final bool isBindingEnabled;
  final BindMode bindMode;
  final double maxCoordinate;
}

typedef ArrowPatch = Map<String, dynamic>;

class BindablePatch {
  const BindablePatch({
    required this.id,
    this.addBoundArrowId,
    this.removeBoundArrowId,
  });

  final String id;
  final String? addBoundArrowId;
  final String? removeBoundArrowId;
}

class ArrowBindingState {
  const ArrowBindingState({
    required this.id,
    required this.startBinding,
    required this.endBinding,
  });

  final String id;
  final FixedPointBinding? startBinding;
  final FixedPointBinding? endBinding;
}

typedef ArrowBindingStatePatch = Map<String, dynamic>;

class BindableRelationState {
  const BindableRelationState({required this.id, required this.boundArrowIds});

  final String id;
  final List<String> boundArrowIds;
}

class BindableRelationPatch {
  const BindableRelationPatch({required this.id, required this.boundArrowIds});

  final String id;
  final List<String> boundArrowIds;
}

class DeriveBindableRelationPatchesForBindingChangeInput {
  const DeriveBindableRelationPatchesForBindingChangeInput({
    required this.arrowId,
    required this.previous,
    required this.next,
    required this.bindables,
  });

  final String arrowId;
  final ArrowBindingState previous;
  final ArrowBindingState next;
  final List<BindableRelationState> bindables;
}

class IdMapEntry {
  const IdMapEntry({required this.from, required this.to});

  final String from;
  final String to;
}

typedef IdMapRecord = Map<String, String>;
typedef IdMapInput = Object;

class ArrowStatePatchWithId {
  const ArrowStatePatchWithId({required this.id, required this.patch});

  final String id;
  final ArrowPatch patch;
}

typedef LifecycleSyncBaseInput = Map<String, dynamic>;
typedef SyncBindingsAfterDuplicationInput = Map<String, dynamic>;
typedef SyncBindingsAfterDeletionInput = Map<String, dynamic>;
typedef SyncBindingsAfterBindablePruneInput = Map<String, dynamic>;

class LifecycleSyncResult {
  const LifecycleSyncResult({
    required this.arrows,
    required this.bindables,
    required this.arrowPatches,
    required this.relationPatches,
    required this.events,
  });

  final List<ArrowState> arrows;
  final List<BindableRelationState> bindables;
  final List<ArrowStatePatchWithId> arrowPatches;
  final List<BindableRelationPatch> relationPatches;
  final List<ArrowEngineEvent> events;
}

typedef BindableLookupRecord = Map<String, BindableState>;
typedef BindableLookupInput = Object;

class SuggestedBinding {
  const SuggestedBinding({
    this.bindableId,
    required this.element,
    this.midPoint,
  });

  final String? bindableId;
  final BindableState element;
  final Point? midPoint;
}

sealed class ArrowEngineEvent {
  const ArrowEngineEvent();

  String get type;
}

class ReorderArrowEvent extends ArrowEngineEvent {
  const ReorderArrowEvent({required this.arrowId, required this.bindableId});

  final String arrowId;
  final String bindableId;

  @override
  String get type => 'reorder-arrow';
}

class BindingBrokenEvent extends ArrowEngineEvent {
  const BindingBrokenEvent({required this.arrowId, required this.edge});

  final String arrowId;
  final ArrowEndpointEdge edge;

  @override
  String get type => 'binding-broken';
}

class EngineResult {
  const EngineResult({
    required this.arrowPatch,
    required this.bindablePatches,
    required this.suggestedBinding,
    required this.events,
  });

  final ArrowPatch arrowPatch;
  final List<BindablePatch> bindablePatches;
  final SuggestedBinding? suggestedBinding;
  final List<ArrowEngineEvent> events;
}

typedef AnchorElementIdsLookupRecord = Map<String, List<String>>;
typedef AnchorElementIdsLookupInput = Object;

class ReduceArrowEngineEventsToOrderInput {
  const ReduceArrowEngineEventsToOrderInput({
    required this.orderedElementIds,
    required this.events,
    this.anchorElementIdsByBindableId,
  });

  final List<String> orderedElementIds;
  final List<ArrowEngineEvent> events;
  final AnchorElementIdsLookupInput? anchorElementIdsByBindableId;
}

class ReduceArrowEngineEventsToOrderResult {
  const ReduceArrowEngineEventsToOrderResult({
    required this.orderedElementIds,
    required this.moved,
    required this.reorderOperations,
    required this.bindingBrokenEvents,
  });

  final List<String> orderedElementIds;
  final bool moved;
  final List<ReorderArrowAboveElementsResult> reorderOperations;
  final List<BindingBrokenEvent> bindingBrokenEvents;
}

class ApplyEngineResultInput {
  const ApplyEngineResultInput({
    required this.arrow,
    required this.bindables,
    required this.result,
    this.orderedElementIds,
    this.anchorElementIdsByBindableId,
  });

  final ArrowState arrow;
  final List<BindableRelationState> bindables;
  final EngineResult result;
  final List<String>? orderedElementIds;
  final AnchorElementIdsLookupInput? anchorElementIdsByBindableId;
}

class ApplyEngineResultValue {
  const ApplyEngineResultValue({
    required this.arrow,
    required this.bindables,
    required this.relationPatches,
    this.orderedElementIds,
    this.orderChanged,
    this.reorderOperations,
    this.bindingBrokenEvents,
  });

  final ArrowState arrow;
  final List<BindableRelationState> bindables;
  final List<BindableRelationPatch> relationPatches;
  final List<String>? orderedElementIds;
  final bool? orderChanged;
  final List<ReorderArrowAboveElementsResult>? reorderOperations;
  final List<BindingBrokenEvent>? bindingBrokenEvents;
}

class ValidationReport {
  const ValidationReport({required this.valid, required this.violations});

  final bool valid;
  final List<String> violations;
}

class EndpointBindingStrategy {
  const EndpointBindingStrategy({
    this.mode,
    this.bindableId,
    this.element,
    this.focusPoint,
  });

  final BindMode? mode;
  final String? bindableId;
  final BindableState? element;
  final Point? focusPoint;
}

class PointUpdate {
  const PointUpdate({required this.index, required this.point});

  final int index;
  final Point point;
}

typedef PointUpdatesRecord = Map<int, Point>;
typedef PointUpdates = Object;

typedef ComputeEndpointDragInput = Map<String, dynamic>;
typedef RefreshEndpointBindingInput = Map<String, dynamic>;
typedef PruneArrowBindingsInput = Map<String, dynamic>;
typedef RecomputeAfterBindableChangeInput = Map<String, dynamic>;
typedef RecomputeBindingsForChangedBindablesInput = Map<String, dynamic>;

class RecomputeBindingsForChangedBindablesResult {
  const RecomputeBindingsForChangedBindablesResult({
    required this.arrows,
    required this.bindables,
    required this.arrowPatches,
    required this.relationPatches,
    required this.events,
  });

  final List<ArrowState> arrows;
  final List<BindableRelationState> bindables;
  final List<ArrowStatePatchWithId> arrowPatches;
  final List<BindableRelationPatch> relationPatches;
  final List<ArrowEngineEvent> events;
}

typedef RecomputeElbowInput = Map<String, dynamic>;

typedef ElbowUpdatePatch = Map<String, dynamic>;
typedef UpdateElbowArrowInput = Map<String, dynamic>;
typedef ComputeElbowResizePatchInput = Map<String, dynamic>;
typedef MoveFixedSegmentToPointInput = Map<String, dynamic>;

class MoveFixedSegmentToPointResult {
  const MoveFixedSegmentToPointResult({
    required this.patch,
    required this.activeSegmentIndex,
    required this.activeSegmentMidPoint,
  });

  final ArrowPatch patch;
  final int? activeSegmentIndex;
  final Point? activeSegmentMidPoint;
}

typedef RepairBindingOnRestoreInput = Map<String, dynamic>;
typedef RepairInvalidUnboundElbowArrowOnRestoreInput = Map<String, dynamic>;
typedef RepairSelfBoundExtremeElbowArrowOnRestoreInput = Map<String, dynamic>;

class FocusPointDescriptor {
  const FocusPointDescriptor({
    required this.edge,
    required this.point,
    required this.binding,
  });

  final ArrowEndpointEdge edge;
  final Point point;
  final FixedPointBinding binding;
}

typedef ListVisibleFocusPointsInput = Map<String, dynamic>;
typedef PickFocusPointInput = Map<String, dynamic>;
typedef PickFocusPointWithOffsetInput = Map<String, dynamic>;

class FocusPointHit {
  const FocusPointHit({required this.edge, required this.pointerOffset});

  final ArrowEndpointEdge? edge;
  final Point pointerOffset;
}

typedef ResizeHandleDirection = Object;
typedef ResizeArrowDirection = String;

typedef ComputeFocusPointDragInput = Map<String, dynamic>;
typedef FinalizeFocusPointDragInput = Map<String, dynamic>;

class ReorderArrowAboveElementsInput {
  const ReorderArrowAboveElementsInput({
    required this.orderedElementIds,
    required this.arrowId,
    required this.anchorElementIds,
  });

  final List<String> orderedElementIds;
  final String arrowId;
  final List<String> anchorElementIds;
}

class ReorderArrowAboveElementsResult {
  const ReorderArrowAboveElementsResult({
    required this.orderedElementIds,
    required this.moved,
    required this.fromIndex,
    required this.toIndex,
  });

  final List<String> orderedElementIds;
  final bool moved;
  final int fromIndex;
  final int toIndex;
}

class ReorderArrowAboveHoveredBindableInput {
  const ReorderArrowAboveHoveredBindableInput({
    required this.orderedElementIds,
    required this.arrowId,
    this.hoveredBindableId,
    this.point,
    this.bindables,
    this.tolerance,
    this.anchorElementIdsByBindableId,
  });

  final List<String> orderedElementIds;
  final String arrowId;
  final String? hoveredBindableId;
  final Point? point;
  final List<BindableState>? bindables;
  final double? tolerance;
  final AnchorElementIdsLookupInput? anchorElementIdsByBindableId;
}

class ReorderArrowAboveHoveredBindableResult
    extends ReorderArrowAboveElementsResult {
  const ReorderArrowAboveHoveredBindableResult({
    required super.orderedElementIds,
    required super.moved,
    required super.fromIndex,
    required super.toIndex,
    required this.hoveredBindableId,
    required this.anchorElementIds,
  });

  final String? hoveredBindableId;
  final List<String> anchorElementIds;
}

const EngineContext defaultEngineContext = EngineContext(
  zoom: 1,
  isBindingEnabled: true,
  bindMode: bindModeOrbit,
  maxCoordinate: 1e6,
);

EngineContext normalizeEngineContext(Map<String, dynamic>? context) {
  final zoom = context?['zoom'];
  final isBindingEnabled = context?['isBindingEnabled'];
  final bindMode = context?['bindMode'];
  final maxCoordinate = context?['maxCoordinate'];

  final normalizedZoom = zoom is num && zoom.isFinite
      ? zoom.toDouble()
      : defaultEngineContext.zoom;
  final normalizedBinding = isBindingEnabled is bool
      ? isBindingEnabled
      : defaultEngineContext.isBindingEnabled;
  final normalizedMode =
      bindMode == bindModeInside ||
          bindMode == bindModeOrbit ||
          bindMode == bindModeSkip
      ? bindMode as String
      : defaultEngineContext.bindMode;
  final normalizedMaxCoordinate = maxCoordinate is num && maxCoordinate.isFinite
      ? maxCoordinate.toDouble()
      : defaultEngineContext.maxCoordinate;

  return EngineContext(
    zoom: normalizedZoom,
    isBindingEnabled: normalizedBinding,
    bindMode: normalizedMode,
    maxCoordinate: normalizedMaxCoordinate,
  );
}

BindableState normalizeBindableState(BindableState bindable) {
  Bounds? normalizedVisibilityBounds;
  final visibilityBounds = bindable.visibilityBounds;
  if (visibilityBounds == null) {
    normalizedVisibilityBounds = null;
  } else if (visibilityBounds.length == 4 &&
      visibilityBounds.every((value) => value.isFinite)) {
    normalizedVisibilityBounds = [
      visibilityBounds[0],
      visibilityBounds[1],
      visibilityBounds[2],
      visibilityBounds[3],
    ];
  }

  return bindable.copyWith(
    shape: canonicalizeBindableShape(bindable.shape),
    backgroundOpaque: bindable.backgroundOpaque ?? true,
    bindingEnabled: bindable.bindingEnabled ?? true,
    interiorHitEnabled: bindable.interiorHitEnabled ?? true,
    visibilityBounds: normalizedVisibilityBounds,
    setVisibilityBounds:
        normalizedVisibilityBounds != null || bindable.visibilityBounds == null,
  );
}

List<BindableState> normalizeBindableStates(List<BindableState> bindables) =>
    bindables.map(normalizeBindableState).toList(growable: false);

bool isFiniteNum(Object? value) => value is num && value.isFinite;

double clampNum(double value, double min, double max) =>
    math.max(min, math.min(max, value));

File: packages/snow_draw_rust_engine/src/draw/elements/types/arrow/core/arrow_types.rs
#![allow(dead_code)]

use std::collections::HashMap;

use serde_json::Value;

use crate::draw::types::draw_point::DrawPoint;
use crate::draw::types::draw_rect::DrawRect;

pub type Point = DrawPoint;
pub type Bounds = DrawRect;
pub type BindMode = &'static str;
pub type Arrowhead = &'static str;
pub type ArrowStrokeStyle = &'static str;
pub type ArrowheadDashMode = &'static str;
pub type ArrowheadFillMode = &'static str;
pub type CanonicalBindableShape = &'static str;
pub type BindableShape = String;
pub type ArrowPatch = serde_json::Map<String, Value>;
pub type ArrowBindingStatePatch = serde_json::Map<String, Value>;
pub type IdMapRecord = HashMap<String, String>;
pub type AnchorElementIdsLookupRecord = HashMap<String, Vec<String>>;

pub const BIND_MODE_INSIDE: BindMode = "inside";
pub const BIND_MODE_ORBIT: BindMode = "orbit";
pub const BIND_MODE_SKIP: BindMode = "skip";

pub const ARROW_ENDPOINT_START: &str = "start";
pub const ARROW_ENDPOINT_END: &str = "end";

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
pub enum ArrowEndpointEdge {
    Start,
    End,
}

impl ArrowEndpointEdge {
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::Start => ARROW_ENDPOINT_START,
            Self::End => ARROW_ENDPOINT_END,
        }
    }
}

pub fn normalize_arrow_endpoint_edge(edge: &str) -> ArrowEndpointEdge {
    match edge {
        "start" | "startBinding" => ArrowEndpointEdge::Start,
        _ => ArrowEndpointEdge::End,
    }
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
pub enum BindableRoundnessType {
    Legacy,
    Proportional,
    Adaptive,
}

#[derive(Clone, Debug, PartialEq)]
pub struct BindableRoundness {
    pub kind: BindableRoundnessType,
    pub value: Option<f64>,
}

impl BindableRoundness {
    pub const fn new(kind: BindableRoundnessType, value: Option<f64>) -> Self {
        Self { kind, value }
    }
}

pub fn canonicalize_bindable_shape(shape: &str) -> String {
    match shape {
        "rect" => "rectangle".to_string(),
        "circle" => "ellipse".to_string(),
        "rhombus" => "diamond".to_string(),
        _ => shape.to_string(),
    }
}

#[derive(Clone, Debug, PartialEq)]
pub struct CurvePathOp {
    pub op: String,
    pub data: Vec<f64>,
}

impl CurvePathOp {
    pub fn new(op: impl Into<String>, data: Vec<f64>) -> Self {
        Self {
            op: op.into(),
            data,
        }
    }
}

#[derive(Clone, Debug, PartialEq)]
pub struct FixedPointBinding {
    pub element_id: String,
    pub fixed_point: Point,
    pub mode: String,
}

impl FixedPointBinding {
    pub fn new(element_id: impl Into<String>, fixed_point: Point, mode: impl Into<String>) -> Self {
        Self {
            element_id: element_id.into(),
            fixed_point,
            mode: mode.into(),
        }
    }
}

#[derive(Clone, Debug, PartialEq)]
pub struct FixedSegment {
    pub start: Point,
    pub end: Point,
    pub index: usize,
}

#[derive(Clone, Debug, PartialEq)]
pub struct BindableState {
    pub id: String,
    pub shape: String,
    pub x: f64,
    pub y: f64,
    pub width: f64,
    pub height: f64,
    pub angle: f64,
    pub stroke_width: f64,
    pub roundness: Option<BindableRoundness>,
    pub z_index: Option<f64>,
    pub background_opaque: Option<bool>,
    pub binding_enabled: Option<bool>,
    pub interior_hit_enabled: Option<bool>,
    pub visibility_bounds: Option<Bounds>,
}

impl BindableState {
    pub fn rect(&self) -> DrawRect {
        DrawRect::new(self.x, self.y, self.x + self.width, self.y + self.height)
    }
}

#[derive(Clone, Debug, PartialEq)]
pub struct ArrowState {
    pub id: String,
    pub x: f64,
    pub y: f64,
    pub width: f64,
    pub height: f64,
    pub points: Vec<Point>,
    pub start_binding: Option<FixedPointBinding>,
    pub end_binding: Option<FixedPointBinding>,
    pub start_arrowhead: Option<String>,
    pub end_arrowhead: Option<String>,
    pub elbowed: bool,
    pub fixed_segments: Option<Vec<FixedSegment>>,
    pub start_is_special: Option<bool>,
    pub end_is_special: Option<bool>,
}

impl ArrowState {
    pub fn rect(&self) -> DrawRect {
        DrawRect::new(self.x, self.y, self.x + self.width, self.y + self.height)
    }
}

#[derive(Clone, Copy, Debug, PartialEq)]
pub struct EngineContext {
    pub zoom: f64,
    pub is_binding_enabled: bool,
    pub bind_mode: &'static str,
    pub max_coordinate: f64,
}

impl Default for EngineContext {
    fn default() -> Self {
        Self {
            zoom: 1.0,
            is_binding_enabled: true,
            bind_mode: BIND_MODE_ORBIT,
            max_coordinate: 1e6,
        }
    }
}

#[derive(Clone, Debug, PartialEq)]
pub struct BindablePatch {
    pub id: String,
    pub add_bound_arrow_id: Option<String>,
    pub remove_bound_arrow_id: Option<String>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct ArrowBindingState {
    pub id: String,
    pub start_binding: Option<FixedPointBinding>,
    pub end_binding: Option<FixedPointBinding>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct BindableRelationState {
    pub id: String,
    pub bound_arrow_ids: Vec<String>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct BindableRelationPatch {
    pub id: String,
    pub bound_arrow_ids: Vec<String>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct ArrowStatePatchWithId {
    pub id: String,
    pub patch: ArrowPatch,
}

#[derive(Clone, Debug, PartialEq)]
pub struct LifecycleSyncResult {
    pub arrows: Vec<ArrowState>,
    pub bindables: Vec<BindableRelationState>,
    pub arrow_patches: Vec<ArrowStatePatchWithId>,
    pub relation_patches: Vec<BindableRelationPatch>,
    pub events: Vec<ArrowEngineEvent>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct SuggestedBinding {
    pub bindable_id: Option<String>,
    pub element: BindableState,
    pub mid_point: Option<Point>,
}

#[derive(Clone, Debug, PartialEq)]
pub enum ArrowEngineEvent {
    ReorderArrow(ReorderArrowEvent),
    BindingBroken(BindingBrokenEvent),
}

#[derive(Clone, Debug, PartialEq)]
pub struct ReorderArrowEvent {
    pub arrow_id: String,
    pub bindable_id: String,
}

#[derive(Clone, Debug, PartialEq)]
pub struct BindingBrokenEvent {
    pub arrow_id: String,
    pub edge: ArrowEndpointEdge,
}

#[derive(Clone, Debug, PartialEq)]
pub struct EngineResult {
    pub arrow_patch: ArrowPatch,
    pub bindable_patches: Vec<BindablePatch>,
    pub suggested_binding: Option<SuggestedBinding>,
    pub events: Vec<ArrowEngineEvent>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct ReduceArrowEngineEventsToOrderInput {
    pub ordered_element_ids: Vec<String>,
    pub events: Vec<ArrowEngineEvent>,
    pub anchor_element_ids_by_bindable_id: Option<AnchorElementIdsLookupRecord>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct ReduceArrowEngineEventsToOrderResult {
    pub ordered_element_ids: Vec<String>,
    pub moved: bool,
    pub reorder_operations: Vec<ReorderArrowAboveElementsResult>,
    pub binding_broken_events: Vec<BindingBrokenEvent>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct ApplyEngineResultInput {
    pub arrow: ArrowState,
    pub bindables: Vec<BindableRelationState>,
    pub result: EngineResult,
    pub ordered_element_ids: Option<Vec<String>>,
    pub anchor_element_ids_by_bindable_id: Option<AnchorElementIdsLookupRecord>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct ApplyEngineResultValue {
    pub arrow: ArrowState,
    pub bindables: Vec<BindableRelationState>,
    pub relation_patches: Vec<BindableRelationPatch>,
    pub ordered_element_ids: Option<Vec<String>>,
    pub order_changed: Option<bool>,
    pub reorder_operations: Option<Vec<ReorderArrowAboveElementsResult>>,
    pub binding_broken_events: Option<Vec<BindingBrokenEvent>>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct ValidationReport {
    pub valid: bool,
    pub violations: Vec<String>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct FocusPointDescriptor {
    pub edge: ArrowEndpointEdge,
    pub point: Point,
}

#[derive(Clone, Debug, PartialEq)]
pub struct FocusPointHit {
    pub edge: Option<ArrowEndpointEdge>,
    pub pointer_offset: Point,
}

#[derive(Clone, Debug, PartialEq)]
pub struct ReorderArrowAboveElementsInput {
    pub ordered_element_ids: Vec<String>,
    pub arrow_id: String,
    pub anchor_element_ids: Vec<String>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct ReorderArrowAboveElementsResult {
    pub ordered_element_ids: Vec<String>,
    pub moved: bool,
    pub from_index: isize,
    pub to_index: isize,
}

#[derive(Clone, Debug, PartialEq)]
pub struct ReorderArrowAboveHoveredBindableInput {
    pub ordered_element_ids: Vec<String>,
    pub arrow_id: String,
    pub hovered_bindable_id: Option<String>,
    pub point: Option<Point>,
    pub bindables: Option<Vec<BindableState>>,
    pub tolerance: Option<f64>,
    pub anchor_element_ids_by_bindable_id: Option<AnchorElementIdsLookupRecord>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct ReorderArrowAboveHoveredBindableResult {
    pub ordered_element_ids: Vec<String>,
    pub moved: bool,
    pub from_index: isize,
    pub to_index: isize,
    pub hovered_bindable_id: Option<String>,
    pub anchor_element_ids: Vec<String>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct NormalizedArrowFromGlobalPoints {
    pub x: f64,
    pub y: f64,
    pub points: Vec<Point>,
    pub width: f64,
    pub height: f64,
}

pub fn is_finite_num(value: f64) -> bool {
    value.is_finite()
}

pub fn clamp_num(value: f64, min: f64, max: f64) -> f64 {
    value.clamp(min, max)
}

pub fn normalize_bindable_state(bindable: &BindableState) -> BindableState {
    let visibility_bounds = bindable.visibility_bounds.and_then(|bounds| {
        let values = [bounds.min_x, bounds.min_y, bounds.max_x, bounds.max_y];
        if values.iter().all(|value| value.is_finite()) {
            Some(bounds)
        } else {
            None
        }
    });

    BindableState {
        id: bindable.id.clone(),
        shape: canonicalize_bindable_shape(&bindable.shape),
        x: bindable.x,
        y: bindable.y,
        width: bindable.width,
        height: bindable.height,
        angle: bindable.angle,
        stroke_width: bindable.stroke_width,
        roundness: bindable.roundness.clone(),
        z_index: bindable.z_index,
        background_opaque: Some(bindable.background_opaque.unwrap_or(true)),
        binding_enabled: Some(bindable.binding_enabled.unwrap_or(true)),
        interior_hit_enabled: Some(bindable.interior_hit_enabled.unwrap_or(true)),
        visibility_bounds,
    }
}

pub fn normalize_bindable_states(bindables: &[BindableState]) -> Vec<BindableState> {
    bindables.iter().map(normalize_bindable_state).collect()
}
